### ISSUE TITLE
azblob: first character of metadata key is automatically capitalized

### ISSUE THREAD

**Issue by ItalyPaleAle** [Azure/azure-sdk-for-go#17850]
### Bug Report

- Package: `github.com/Azure/azure-sdk-for-go/sdk/storage/azblob@v0.4.0`
- output of `go version`: `go version go1.18.1 linux/arm64 `

When setting metadata (e.g. while uploading a blob with `BlockBlobClient.UploadStream`), the first character of the metadata’s key is automatically capitalized.

For example, if you set this object as metadata:

```go
metadata := make(map[string]string)
metadata["foo"] = "bar"
metadata["hello"] = "world"
```

The created blob has metadata:

```
Foo = bar
Hello = world
```

The capitalization should not happen automatically (and I don’t believe the old track1 SDK had this behavior)

**Comment by jhendrixMSFT** [comment]
This is due to use of `http.Header.Set()` which canonicalizes the HTTP header values.

https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/storage/azblob/zz_generated_blockblob_client.go#L104

While we could change the generated code to set the value directly, it would be ideal if there was a way for client libraries to control the behavior.  Unless we never want canonicalization.

**Comment by jhendrixMSFT** [comment]
The codegen change will be in `preview.41` of the code generator.

**Comment by jhendrixMSFT** [comment]
Note that while this fixes _sending_ of metadata keys, the _receiving_ of keys is still broken due to the default transport canonicalizing the response.

**Comment by ItalyPaleAle** [comment]
Hi! Is there any update on when this issue will be fixed?

**Comment by jhendrixMSFT** [comment]
The canonicalization of received HTTP headers comes from the HTTP stack in the standard library.  At present, there's no way to disable this behavior.  This happens [here](https://github.com/golang/go/blob/master/src/net/textproto/reader.go#L515).

See also https://github.com/golang/go/issues/37834

**Comment by ItalyPaleAle** [comment]
@jhendrixMSFT thanks for sharing that. Why does the "track1" SDK not have this issue, then? I don't believe I've seen headers "canonicalized" there, unless I remember incorrectly.

**Comment by jhendrixMSFT** [comment]
We applied the same fix when sending headers.  Receiving them is still a problem though.

**Comment by ItalyPaleAle** [comment]
Ok, track1 is working fine for Dapr, so perhaps I can try even if the receiving side doesn't have that fix.

Is the fix on the sending side published now?

**Comment by jhendrixMSFT** [comment]
Not yet, I'll follow up on when this will happen.

Also note that we're evaluating some significant refactoring of the public surface area for `azblob`.  See [here](https://apiview.dev/Assemblies/Review/858fef15e4d24307a1372285fd35ce17) for what it might look like.  No guarantees that it will take this exact shape.  Feel free to send me any feedback on it offline.

**Comment by ItalyPaleAle** [comment]
Ok. We will wait. Link isn't working for me btw.

**Comment by jhendrixMSFT** [comment]
Must be a permissions issue (the link was to an API review).  I updated the link to just point to my fork where the work resides.

**Comment by onionhammer** [comment]
This seems to be the source for a lot of downstream blockers; any update?

**Comment by jhendrixMSFT** [comment]
Our updated `azblob` should be released in the upcoming weeks (ideally week of 8/29, but it might slip to the first week of September).

**Comment by jhendrixMSFT** [comment]
Fixed in `azblob@v0.5.0`.
